x-pack/legacy/plugins/code/server/git_operations.ts

```diff
@@ -443,210 +292,192 @@ export class GitOperations {
-    for (const oid of commit.parents()) {
-      const parentCommit = await repo.getCommit(oid);
-      if (parentCommit) {
-        const entry = await parentCommit.getEntry(path);
-        if (entry) {
-          return (await entry.getBlob()).content().toString('utf8');
-        }
-      }
+  private diffKind(diff: DiffResultTextFile) {
+    let kind: DiffKind = DiffKind.MODIFIED;
+    if (diff.changes === diff.insertions) {
+      kind = DiffKind.ADDED;
+    } else if (diff.changes === diff.deletions) {
+      kind = DiffKind.DELETED;
+    } else if (diff.rename) {
+      kind = DiffKind.RENAMED;
     }
-    return '';
+    return kind;
   }
 
-  private async getModifiedCode(commit: NodeGitCommit, path: string) {
-    const entry = await commit.getEntry(path);
-    return (await entry.getBlob()).content().toString('utf8');
+  private async getOriginCode(git: SimpleGit, commit: CommitInfo, path: string) {
+    const buffer: Buffer = await git.binaryCatFile(['blob', `${commit.id}~1:${path}`]);
+    return buffer.toString('utf8');
   }
 
-  private async findCommit(repo: Repository, oid: string): Promise<NodeGitCommit | null> {
-    try {
-      return repo.getCommit(Oid.fromString(oid));
-    } catch (e) {
-      return null;
-    }
+  private async getModifiedCode(git: SimpleGit, commit: CommitInfo, path: string) {
+    const buffer: Buffer = await git.binaryCatFile(['blob', `${commit.id}:${path}`]);
+    return buffer.toString('utf8');
   }
 
   public async getBranchAndTags(repoUri: string): Promise<ReferenceInfo[]> {
-    const gitdir = this.repoDir(repoUri);
-    const remoteBranches = await GitPrime.listBranches({ gitdir, remote: 'origin' });
-    const results: ReferenceInfo[] = [];
-    for (const name of remoteBranches) {
-      const reference = `refs/remotes/origin/${name}`;
-      const commit = await this.getCommitInfo(repoUri, reference);
-      if (commit) {
-        results.push({
-          name,
-          reference,
-          type: ReferenceType.REMOTE_BRANCH,
-          commit,
-        });
-      }
-    }
-    const tags = await GitPrime.listTags({ gitdir });
-    for (const name of tags) {
-      const reference = `refs/tags/${name}`;
-      const commit = await this.getCommitInfo(repoUri, reference);
-      if (commit) {
-        results.push({
-          name,
-          reference,
-          type: ReferenceType.TAG,
-          commit,
-        });
+    const format = {
+      name: '%(refname:short)',
+      reference: '%(refname)',
+      type: '%(objecttype)',
+      commit: {
+        updated: '%(*authordate)',
+        message: '%(*contents)',
+        committer: '%(*committername)',
+        author: '%(*authorname)',
+        id: '%(*objectname)',
+        parents: '%(*parent)',
+        treeId: '%(*tree)',
+      },
+    };
+    const parser = new FormatParser(format);
+    const git = await this.openGit(repoUri);
+    const result = await git.raw([
```

Comment:
Hopefully, there is no security issue here. Might worth to let Alex take a post review after this PR has been merged.

Reply:
It's a simple git command, see doc here https://git-scm.com/docs/git-for-each-ref.

Reply:
it's not the command itself which is risky, we want to make sure the command can't be injected,
e.g.

`git log xxxx` is ok, but we don't want it to be inject to become `git log xxx;rm -rf`